### PR TITLE
[megatron] feat: fused kernel suppport for new model engine

### PR DESCRIPTION
**Pull request by HollowMan6**
### What does this PR do?

Add `fused_forward_no_padding` method so that model engine can also support fused kernels.

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: ...
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

Tested on 8*A800 GPUs with triton version 3.5.1 using DAPO-Math-17k dataset with DAPO and using Moonlight-16B-A3B, the training curves are closely matched by either disabling or enabling fused kernels, and even enabling fused kernels has better rollout training mismatch level.

image
image
image
image

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [X] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [X] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/workers/engine/megatron/transformer_impl.py [hunk 1]
```diff
@@ -665,7 +676,41 @@ def forward_step(self, batch_iter: Iterator[TensorDict], model, postprocess_micr
...
+            output = fused_forward_fn(
+                model=model,
+                input_ids=input_ids,
+                labels=label,
+                multi_modal_inputs=multi_modal_inputs,
+                temperature=temperature_value,
+                calculate_entropy=calculate_entropy,
+                pad_token_id=self.model_config.tokenizer.pad_token_id,
+            )
+            return output, partial(postprocess_micro_batch_func, data=batch)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/engine/megatron/transformer_impl.py:706]: fixed

File: verl/workers/engine/megatron/transformer_impl.py [hunk 2]
```diff
@@ -694,7 +739,7 @@ def logits_processor(logits, label, temperature):
             ret["log_probs"] = log_probs
             return ret
 
-        logits_processor_args = {"label": label, "temperature": temperature, "loss_mask": loss_mask}
+        logits_processor_args = {"label": label, "temperature": temperature}
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/engine/megatron/transformer_impl.py]: fixed

File: verl/workers/engine/megatron/transformer_impl.py [hunk 3]
```diff
@@ -648,14 +666,7 @@ def forward_step(self, batch_iter: Iterator[TensorDict], model, postprocess_micr
...
         multi_modal_inputs = model_inputs["multi_modal_inputs"]
-        loss_mask = model_inputs["loss_mask"]
-
-        if not isinstance(temperature, torch.Tensor):
-            temperature = torch.tensor([temperature] * input_ids.shape[0], device=input_ids.device)
-
-        temperature = temperature.to(torch.float32)
-        assert temperature.shape[0] == input_ids.shape[0]
-        temperature = verl_F.expand_as_nested(temperature, input_ids)  # (bsz, j1)
+        # loss_mask = model_inputs["loss_mask"]
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/engine/megatron/transformer_impl.py]: fixed

**Comment by ISEEKYAN** [comment]
hi @HollowMan6, we found the fused kernel is broken with the speed after triton3.3 about 3 months ago, since then we don't maintain recommend the fused kernel. Do you observe a speed down after enabling this fused kernel.

As a option, my colleagues have rewrite the fused kernel and integrated it into the MCore, see https://github.com/NVIDIA/Megatron-LM/pull/3076 and https://github.com/NVIDIA/Megatron-LM/pull/2739. But unfortunately now this kernel only supports Blackwell, not even for Hopper. And we have another PR to adopt the mcore version of fused_kernel at https://github.com/verl-project/verl/pull/5130.

I think the best way to use this kernel is to enable it through `transformer_config`, but this can't help Hopper and Ampere users. 
The existing way is OK but I want you to check if the speed is slow down significantly, thanks!

**Review by ISEEKYAN — APPROVED** [review]
The code changes are OK, it only depends on the fused kernel perf

**Comment by HollowMan6** [comment]
Hi @ISEEKYAN ! Thanks for your reviewing and that's interesting to know! Actually, I didn't find a speed degradation on A800 when training Moonlight-16B-A3B with triton 3.5.1, please check the last figure in the PR description, in contrast, it improves the speed a lot (step time around 340 -> around 270)!

If you were benchmarking this on Hopper GPUs 3 months ago and found the speed degradation, actually there's a PR to improve it about 1 and a half months ago and the speed up was also very obvious: https://github.com/verl-project/verl/pull/4576 I've also tested it with triton 3.5.0 together with this support fix https://github.com/verl-project/verl/pull/4676 on L20x and also didn't notice any performance degradation at that time as well.

Unfortunately I don't have blackwell GPUs at hand and can't really help try out/test the new kernel, but since I still observe performance improvement when enabling fused kernel on verl side for now and #4576 also does some improvement on Hopper side, maybe we can still go back and recommend the fused kernel (though for now I no longer have Hopper GPUs on my side and can't really tell how good it is there).

Since I didn't make fused kernel the default option for model engine in this PR, I think it's safe and also good to ask other people to try this out and check the performance benefit, so I'll go ahead and merge this one.

**Comment by ISEEKYAN** [comment]
thanks @HollowMan6, good to know that the great works in https://github.com/verl-project/verl/pull/4576
